### ISSUE TITLE
Loading state design

### ISSUE THREAD

**Issue by Poincare** [apollostack/apollo-client#466]
As discussed in #342, we want Apollo Client to handle most of the logic around loading states and `returnPartialData` rather than passing on this responsibility to view layer implementations such as `angular-apollo` or `react-apollo`. To do this, I am proposing the following design and would be happy to hear feedback.

## Design

We will add a `loading` field to `ApolloQueryResult` which will be a boolean value that describes whether the data requested by a particular is currently in the loading state. For example:

```
const handle = client.watchQuery(gql`
  query {
    //whatever
  }
`);
handle.subscribe({
  next(result) {
    if (result.loading) {
     // this result is still "loading" and may be some of the data that fulfills the original query
   }
  }
});
```

Doing this allows us to support `returnPartialData` effectively. Until the view layer receives a result with a `loading` field, the view layer should default to the `loading` state (i.e. `loading: true`). This allows us to maintain the existing behavior for the queries which do not return partial data.

**Comment by stubailo** [comment]
This looks great!

**Comment by Poincare** [comment]
This was merged with #467
